File: ios/HybridTextRecognitionResult.swift
import Foundation
import NitroModules

final class HybridTextRecognitionResult: HybridTextRecognitionResultSpec {
  private var storedText: String
  private var storedBlocks: [RecognizedTextBlock]
  private var disposed = false

  init(output: TextRecognitionOutput) {
    self.storedText = output.text
    self.storedBlocks = output.blocks
    super.init()
  }

  var text: String { storedText }

  var blockCount: Double { Double(storedBlocks.count) }

  var blocks: [RecognizedTextBlock] { storedBlocks }

  func blockAt(index: Double) throws -> RecognizedTextBlock {
    try ensureNotDisposed()
    let i = Int(index)
    guard i >= 0, i < storedBlocks.count else {
      throw RuntimeError("blockAt index \(i) out of range (blockCount=\(storedBlocks.count)).")
    }
    return storedBlocks[i]
  }

  var memorySize: Int {
    if disposed { return HybridMemorySize.overhead }
    return storedText.utf8.count
      + storedBlocks.reduce(0) { partial, block in
        partial
          + block.text.utf8.count
          + (block.cornerPoints?.count ?? 0) * 16
          + block.lines.reduce(0) { linePartial, line in
            linePartial
              + line.text.utf8.count
              + (line.cornerPoints?.count ?? 0) * 16
              + (line.candidates?.reduce(0) { $0 + $1.text.utf8.count } ?? 0)
          }
      }
      + HybridMemorySize.overhead
  }

  func dispose() {
    disposed = true
    storedText = ""
    storedBlocks = []
  }

  private func ensureNotDisposed() throws {
    if disposed {
      throw RuntimeError(
        "TextRecognitionResult already disposed. Read text/blocks before dispose()."
      )
    }
  }
}

File: android/src/main/java/com/margelo/nitro/nitrovisionkit/HybridTextRecognitionResult.kt
package com.margelo.nitro.nitrovisionkit

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip

@Keep
@DoNotStrip
class HybridTextRecognitionResult internal constructor(
  output: TextRecognitionOutput,
) : HybridTextRecognitionResultSpec() {

  private var storedText: String = output.text
  private var storedBlocks: Array<RecognizedTextBlock> = output.blocks
  private var disposed = false

  override val text: String
    get() {
      ensureNotDisposed()
      return storedText
    }

  override val blockCount: Double
    get() {
      ensureNotDisposed()
      return storedBlocks.size.toDouble()
    }

  override val blocks: Array<RecognizedTextBlock>
    get() {
      ensureNotDisposed()
      return storedBlocks
    }

  override fun blockAt(index: Double): RecognizedTextBlock {
    ensureNotDisposed()
    val i = index.toInt()
    if (i < 0 || i >= storedBlocks.size) {
      throw RuntimeException("blockAt index $i out of range (blockCount=${storedBlocks.size}).")
    }
    return storedBlocks[i]
  }

  override val memorySize: Long
    get() {
      if (disposed) return HybridMemorySize.OVERHEAD
      var size = storedText.length.toLong() + HybridMemorySize.OVERHEAD
      for (block in storedBlocks) {
        size += block.text.length
        size += (block.cornerPoints?.size ?: 0) * 16L
        for (line in block.lines) {
          size += line.text.length
          size += (line.cornerPoints?.size ?: 0) * 16L
          line.candidates?.forEach { size += it.text.length }
        }
      }
      return size
    }

  override fun dispose() {
    disposed = true
    storedText = ""
    storedBlocks = emptyArray()
  }

  private fun ensureNotDisposed() {
    if (disposed) {
      throw RuntimeException(
        "TextRecognitionResult already disposed. Read text/blocks before dispose().",
      )
    }
  }
}
